docs/rfc/0001_ctrlkit.md

```diff
@@ -0,0 +1,325 @@
+
+	RunNextScheduledJob(ctx context.Context, logger logr.Logger) (ctrl.Result, error)
+
+	UpdateCronJobStatus(ctx context.Context, logger logr.Logger) (ctrl.Result, error)
+}
+
+type CronJobControllerManager struct {
+	state  CronJobControllerManagerState
+	impl   CronJobControllerManagerImpl
+	logger logr.Logger
+}
+
+func (m *CronJobControllerManager) ListActiveJobsAndUpdateStatus() ctrlkit.ReconcileAction {
+	return ctrlkit.WrapAction("ListActiveJobsAndUpdateStatus", func(ctx context.Context) (ctrl.Result, error) {
+		logger := m.logger.WithValues("action", "ListActiveJobsAndUpdateStatus")
+
+		jobs, err := m.state.GetJobs(ctx)
+		if err != nil {
+			return ctrlkit.RequeueIfError(err)
+		}
+
+		defer m.impl.AfterActionRun("ListActiveJobsAndUpdateStatus", ctx, logger)
+		m.impl.BeforeActionRun("ListActiveJobsAndUpdateStatus", ctx, logger)
+
+		return m.impl.ListActiveJobsAndUpdateStatus(ctx, logger, jobs)
+	})
+}
+
+...
+```
+
+The developer is responsible for writing an implementation of the `CronJobControllerManagerImpl` to implement the actions.
+
+### **Library**
+
+The library should contain three main parts:
+* Abstraction of actions `ReconcileAction`.
+* Functions for organizing the workflow `Join`, `Sequential`, `Timeout`, and `If`.
+* Functions and errors for continuing or interrupting the workflow.
+
+```go
+type ReconcileAction interface {
+    Description() string
+	Run(ctx context.Context) (ctrl.Result, error)
+}
+```
+
+**Workflow organizing functions**
+
+Function | Example | Effect 
+:---: | :---: | :---: 
+`Join(actions...)`, `ParallelJoin(actions...)`, `JoinOrdered(actions...)` | `Join(Action_A, Action_B)` | Run all actions joined, and return the results joint from each action (while keep the semantic of the results). `ParallelJoin` provides parallism besides the `Join` semantic. `JoinOrdered` guarantees the execution order.
+`Sequential(actions...)` | `Sequential(Action_A, Action_B)` | Run all actions one-by-one. If any of the actions interrupts the workflow, left actions will not be executed and the result of that action will be returned.
+`Timeout(timeout, action)` | `Timeout(5 * time.Second, Action_A)` | Abort the action if specified time is used.
+`If(condition, action)` | `If(x == 1, Action_A)` | Action's only valid when condition is true.
+
+`Sequential` defines a sequential flow of actions, and `Join` defines a split-join flow.
+
+An example workflow with the actions defined above:
+
+```go
+// Run these actions and doesn't care the order, and join the results.
+ctrlkit.Join(
+	// Update the status of CronJob as always.
+	mgr.ListActiveJobsAndUpdateStatus(),
+	// Clean the old completed/failed jobs accroding to the limits.
+	mgr.CleanUpOldJobsExceedsHistoryLimits(),
+	// Try to run the next scheduled job when not suspended, otherwise do nothing.
+	ctrlkit.If(cronJob.Spec.Suspend == nil || *cronJob.Spec.Suspend, mgr.RunNextScheduledJob()),
```

Comment:
Can we do `If(condition, ParallelJoin(Timeout(5, some_action), another_action))`?

Reply:
Sure. In graph if would be like

```plain
                            -> Timeout(5, some_action) -> 
                          /                               \
-> Eval Condition - Yes ->                                  -> ...
                  |       \                               /
                  |         ->     another_action      ->
                  |
                   \
                     - No -> ...
```